src/python/pants/core/goals/deploy.py

```diff
@@ -199,12 +218,28 @@ async def run_deploy(console: Console, deploy_subsystem: DeploySubsystem) -> Dep
         for field_set in target_roots_to_deploy_field_sets.field_sets
     )
 
-    publish_targets = (
+    deploy_publish_dependencies = await MultiGet(
+        Get(
+            Targets,
+            UnparsedAddressInputs,
+            field_set.publish_dependencies.to_unparsed_address_inputs(),
+        )
+        for field_set in target_roots_to_deploy_field_sets.field_sets
+    )
+    specified_publish_targets = (
+        set(chain.from_iterable(deploy_publish_dependencies))
+        if deploy_subsystem.publish_dependencies
+        else set()
+    )
+
+    inferred_publish_targets = (
         set(chain.from_iterable([deploy.publish_dependencies for deploy in deploy_processes]))
         if deploy_subsystem.publish_dependencies
         else set()
     )
 
+    publish_targets = inferred_publish_targets | specified_publish_targets
```

Comment:
Right now we're using `--no-experimental-deploy-publish-dependencies` in CI, because we publish docker images in a separate step, so we don't need to do it again in `deploy` step. So yes, if these 2 options are going to be used together, we would need the set intersection

Reply:
I think I've missed something. This is how I thought of it
- `inferred_publish_targets` : targets a backend infers should be published. For example, Helm can infer Docker images that need to be published before this deployment
- `specified_publish_targets` : manually specified dependencies that should be published before this deployment

Like for normal dependencies, backends can infer dependencies and users can manually add some. The result should be both of these (set union).

I would think that `--no-experimental-deploy-publish-dependencies` would turn both of these off.

Reply:
I think it's very confusing because `deploy_subsystem.publish_dependencies` and `target.publish_dependencies` have the same names but different meaning. It would be much better to change the field name to `dependencies_to_publish` or `runtime_deploy_dependencies` (similar to `runtime_package_dependencies` because these are dependencies need to be deployed and present in runtime)

Reply:
off topic: Personally I don't like the name `runtime_package_dependencies`, and I would prefer `dependencies_to_package` and `dependencies_to_publish` instead